From 50159fe297ca46145558b60a0354701e3ce1bb54 Mon Sep 17 00:00:00 2001
From: Johannes Schindelin <johannes.schindelin@gmx.de>
Date: Wed, 14 Mar 2018 17:04:19 +0100
Subject: [PATCH 51/N] Fix incorrect path conversion (trailing slash)

This commit reverts part of 8eae854a91 (Special case for converting root
directory to have training slash, 2016-03-01) which not only introduced
a funny typo, but also introduced overzealous behavior where the
path_conv class would keep trailing slashes in *waaaaaaay* too many
cases. Unfortunately, the original commit's message as well as the code
comment it introduced are seriously deficient in clarity and are only
excelling in their terseness that makes them pretty useless.

So the best we can do is assume that the change that caused the trailing
slashes to be kept in pretty much all cases (i.e. failing to normalize
the path `Y:\directory\` to `Y:\directory`) was introduced by mistake.

This mistake has serious repercussions, so the above rant has merit.
On FAT file systems, the normalized path is used to fake inode numbers.
As a result, `Y:\directory\` and `Y:\directory` have different inode
numbers!!!

One consequence of this bug has been reported on the Git for Windows:
the `find` and `rm` commands can error out on FAT file systems with very
confusing "No such file or directory" errors, for no good reason.

Vasil Minkov even pointed out that this bug had been fixed in Cygwin as
early as 1997... and we reintroduced it in MSYS2.

This fixes https://github.com/git-for-windows/git/issues/1497 and
https://github.com/git-for-windows/git/issues/1536.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>
---
 winsup/cygwin/path.cc | 18 +++++++++---------
 1 file changed, 9 insertions(+), 9 deletions(-)

diff --git a/winsup/cygwin/path.cc b/winsup/cygwin/path.cc
index 52a95bc..d67a127 100644
--- a/winsup/cygwin/path.cc
+++ b/winsup/cygwin/path.cc
@@ -1226,17 +1226,17 @@ path_conv::check (const char *src, unsigned opt,
 		cfree (wide_path);
 	      wide_path = NULL;
 	    }
-	}
 
-      if (need_directory)
-	{
-	  size_t n = strlen (this->path);
-	  /* Do not add trailing \ to UNC device names like \\.\a: */
-	  if (this->path[n - 1] != '\\' &&
-	      (strncmp (this->path, "\\\\.\\", 4) != 0))
+	  if (need_directory)
 	    {
-	      this->modifiable_path ()[n] = '\\';
-	      this->modifiable_path ()[n + 1] = '\0';
+	      size_t n = strlen (this->path);
+	      /* Do not add trailing \ to UNC device names like \\.\a: */
+	      if (this->path[n - 1] != '\\' &&
+		  (strncmp (this->path, "\\\\.\\", 4) != 0))
+		{
+		  this->modifiable_path ()[n] = '\\';
+		  this->modifiable_path ()[n + 1] = '\0';
+		}
 	    }
 	}
 
-- 
2.9.0

